src/backends/matrix_state_flat.hpp

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <memory/memory.hpp>
+#include <util/span.hpp>
+#include <util/rangeutil.hpp>
+
+#include "gpu_kernels.hpp"
+
+namespace nest {
+namespace mc {
+namespace gpu {
+
+/// matrix state
+template <typename T, typename I>
+struct matrix_state_flat {
+    using value_type = T;
+    using size_type = I;
+
+    using array  = memory::device_vector<value_type>;
+    using iarray = memory::device_vector<size_type>;
+
+    using view = typename array::view_type;
+    using const_view = typename array::const_view_type;
+
+    iarray parent_index;
+    iarray cell_index;
+
+    array d;     // [μS]
+    array u;     // [μS]
+    array rhs;   // [nA]
+
+    array cv_capacitance;      // [pF]
+    array face_conductance;    // [μS]
+
+    // the invariant part of the matrix diagonal
+    array invariant_d;         // [μS]
+
+    // interface for exposing the solution to the outside world
+    view solution;
+
+    matrix_state_flat() = default;
+
+    matrix_state_flat(const std::vector<size_type>& p,
+                 const std::vector<size_type>& cell_index,
+                 const std::vector<value_type>& cv_cap,
+                 const std::vector<value_type>& face_cond):
+        parent_index(memory::make_const_view(p)),
+        cell_index(memory::make_const_view(cell_index)),
+        d(p.size(), impl::npos<T>()),
+        u(p.size()),
+        rhs(p.size(), impl::npos<T>()),
+        cv_capacitance(memory::make_const_view(cv_cap))
+    {
```

Comment:
But you won't like the solution! Instead of having all the files with `_gpu` and `_multicore` in their name, I made separate paths. Things were already messy before this suggested refactoring.

Now all the GPU-specific code is in `backends/gpu`. This doesn't exactly reflect the `nest::mc::gpu` and `nest::mc::multicore` namespace structure... maybe we could remove the `multicore` path?